### ISSUE TITLE
Overriding annotation/usage keywords at point of use

### ISSUE THREAD

**Issue by handrews** [json-schema-org/json-schema-spec#98]
### Use Case and Motivation

This is one of several proposals to come out of the analysis of the "ban additional properties mode" vs `$merge`/`$patch` debate and [the various use cases that fed into that](https://groups.google.com/forum/#!topic/json-schema/dDkozAUTxRk).

One common use case identified was specifying information about a type's _usage_ at the point of use rather than as part of the initial type definition.  In some cases, it is reasonable to provide some generic usage information in the type definition which serves as default usage documentation, but needs to be overridden in some uses with more specific information.

Additionally, `default` cannot meaningfully appear in multiple branches of logical keywords such as `allOf`, as there is no sensible way to choose which value is correct for the specific usage.  Conflicting validations in an `allOf` simply fail the `allOf`, but conflicting `default` values are just unusable, independent of validation.
#### Prior experience

This particular use case has been a major pain point on both large-scale JSON Schema projects with which I have been involved.

One implemented `$merge`, which was very effective but (as several people have noted, and as I now agree) it is too powerful as it (and `$patch`) can literally convert any schema into an arbitrarily different schema, or even into a non-schema JSON value.  In this project, `$merge` was used to override default values for `default`, `title`, `description`, and `readOnly`, or disambiguate among several possible values for those keywords.

The other project ended up redefining many types that should be sharing a definition.  In some cases, nearly every schema has had to redefine a particular sub-schema just to change the title and description.
#### Preserving independent validation

In order to preserve the self-contained nature of validation, usage overriding _only_ applies to the keywords in the "Metadata keywords" section of the validation specification (`title`, `description`, and `default`) plus usage-oriented hyper-schema keywords (`readOnly` and `pathStart`).

Breaking self-contained validation has been the main objection to other proposals in this area, either through overly powerful keywords or changing validation modes in ways which are not reflected in the schema itself.  This is the only proposal to date that solves this problem with no impact on validation whatsoever.
### Existing workarounds

It is possible to use an `allOf` with a reference to the type schema and a schema that includes only the point-of-use annotations.  This is tolerable when reading the schema, but is not good for documentation generation.  Each schema within an `allOf` should be self-contained, which separates the documentation from the thing it's documenting.  Additionally, if default annotation fields have been filled out, a documentation tool that attempts to extract and combine annotation data from all `allOf` branches will just produce conflicting documentation.

Finally, `allOf` is nonsensical when combining multiple schemas using `default`, `readOnly`, or `pathStart`, as there is no reasonable way to choose which value to use.
### The proposal:  `$use`

The keyword `$use` would be an annotation/hyperschema keyword indicating that a type (via a `$ref`) is being used with additional annotation and/or hypermedia for this specific use.

`$use` is an object with two required properties and one optional property; the format for the required properties is borrowed from the `$merge` and `$patch` proposals:
- `source` MUST be a schema, and will nearly always be a `$ref` as there is no need for `$use` if you have the literal schema defined in line.  Required.
- `with` should be applied to the resolved source as an `application/merge-patch+json` instance, which MUST NOT affect any JSON Schema-defined keyword that is not explicitly allowed by the proposal.  Required.
  - Non-keyword property names _are_ allowed in order to facilitate JSON Schema extensions that can opt to apply `$use` to their extensions or not as they choose.  Required.
#### JSON Merge Patch considerations

Declaring the `with` object to follow JSON Merge Patch semantics allows implementations to use existing libraries, and frees us up from needing to debate yet another schema combination format.

However, a notable limitation of JSON Merge Patch is that you cannot overwrite something to `null`, as specifying a `null` value instead deletes the key entirely.  A workaround for this that is consistent with existing JSON Schema behavior is to define `null` somewhere (perhaps right in the `$use` object since it does not forbid additional properties) and then `$ref` it.  We would need to declare that `$use` is processed before `$ref`.  Which is fine because `$ref` often needs lazy evaluation due to circular references anyway. 

Elsewhere, there's been an assertion that `$ref` must always resolve to a schema.  This would require breaking that assumption.  Since the assumption was not present in Draft 04 (in which JSON Reference was an entirely separate specification), I think this isn't too unreasonable.  A compromise could be that it can reference either a schema or null.

If $ref-ing null doesn't gain acceptance, another option would be an additional keyword in the `$use` object listing JSON pointers to properties that should be set to null.
#### "$" in name

I chose `$use` rather than `use` on the grounds that keywords that manipulate the schema itself should stand out (I will be filing a more comprehensive issue on the use of `$` later as it comes up in yet another not-yet-filed proposal, plus there is an open issue to consider `$id` in place of `id`).
#### Example

``` JSON
{
    "type": "object"
    "properties:" {
        "interestTimestamp": {
            "$use": {
                "source": {"$ref": "#/definitions/specialTimestamp"},
                "with": {
                    "title": "Last Event of Interest",
                    "description": "The last time that something interesting happened.  Cannot be directly updated through the API.",
                    "readOnly": true,
                }
            }
        }
    },
    "definitions": {
        "specialTimestamp": {…}
    }
}
```

This shows that the type can be concisely defined in a shared location, with a clear usage object applied to it.
#### Meta-schema considerations

This could be added to the meta-schema by moving most of the keyword definitions into the meta-schema's `definitions` section grouped by type of keyword, and defining `with` as a schema minus the forbidden keywords.

**Comment by epoberezkin** [comment]
I miss the point completely - why do you need a special construct to add meta keywords?

Why not just:

```json
{
  "title": "foo",
  "$ref": "bar"
}
```

as title etc. are ignored anyway?

Or if you want you can:

```json
{
  "title": "foo",
  "allOf": [ { "$ref": "bar" } ]
}
```
$merge/$patch and banUnknownProperties idea solve additionalProperties issue. $merge/$patch also solve schema extension issue for recursive schemas (meta-schema is just an example, recursive schemas are very common). I agree with the objections against banUnknownProperties. The argument that $merge/$patch is powerful and allows to change the schema is an argument to use it, rather than to avoid.

Leaving aside comparisons to other ideas, what problem that can't be solved already does this proposal solve?

**Comment by handrews** [comment]
@epoberezkin did you read the linked email thread?  It explains exactly how this proposal came about and how it relates to the other issues mentioned, and why it is preferable to $merge/$patch.  I'm not going to rewrite it all here, although I can copy in key points if people think it will help.

Issue #15 is the one for $merge/$patch, and it has numerous points related to this.

@tadhgpearson agreed with the need for this use case in [this comment](https://github.com/json-schema-org/json-schema-spec/issues/15#issuecomment-245045116).

@seagreen and @awwright cover a lot of the concerns in comments after that.

> $merge/$patch also solve schema extension issue for recursive schemas 

This issue is not about recursive schemas, please keep the use cases separate.  **If** we happen to adopt $merge/$patch for other reasons, this issue can be closed as moot, but unless that happens (and I've seen no indication that it will), this issue is *only* about the use case I identified here.

In another issue you suggest that I try out $merge because you think it will grow on me.  That's hilarious- as you can see in #15, my previous project made use of $merge well over 200 times in the published schemas alone (and more in internal schemas).  This proposal is based on the experience that despite a great deal of `"additionalProperties": false` use, we *never* needed $merge to get around that.

**Comment by epoberezkin** [comment]
@handrews I've read the thread.

It does not answer the question why annotations need re-writing and why simply adding them to $ref as in my examples is not sufficient.

I think the idea of re-writing anything follows from treating $ref as equivalent to inclusion. I think it is a big mistake to treat it in this way, because, as I wrote before:

1. There are recursive cases when inclusion is impossible (and some are mutually recursive, so it's not visible from the URI that they are recursive)
2. Reference resolution is based on the source schema and when you include a fragment it won't have it's own ID, nevertheless it should resolve refs based on the ID of the source schema.

Consider this simplified example:

```json
{
  "id": "schema1",
  "allOf": [ { "$ref": "schema2#/definitions/foo" } ]
}
```

```json
{
  "id": "schema2",
  "definitions": {
    "foo": { "$ref": "#/definitions/bar" },
    "bar": { "type": "string" }
  }
}
```

That is a very common use case. If you treat "$ref" as inclusion then `"#/definitions/bar"` would mean definitions in `schema1` and it won't resolve.

This example is equivalent to the [test case in JSON-Schema-Test-Suite](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/develop/tests/draft4/refRemote.json#L35) - if $ref were equivalent to inclusion this test would have fail.

There are use cases when allowing inclusion would be beneficial - it would support polymorphism (same subschemas doing different validation depending on where they are included). If we think it can be beneficial then we may consider an additional keyword that does inclusion (or maybe "$merge" without "with" would do just nicely).

**Comment by handrews** [comment]
> why simply adding them to $ref as in my examples is not sufficient.

Because the only field allowed in a $ref object is $ref.  It references, it doesn't merge.

**Comment by epoberezkin** [comment]
As I wrote above, you can:

```json
{
  "title": "foo",
  "allOf": [ { "$ref": "bar" } ]
}
```

Why is it not sufficient?

**Comment by handrews** [comment]
> There are recursive cases when inclusion is impossible (and some are mutually recursive, so it's not visible from the URI that they are recursive)

If you want to dig into the behavior of $ref, please do so in issue #66 rather than clogging up this issue with topics that are at best tangential to the proposal at hand.

If there is a specific issue where you see this as a problem with $use, please explain it in terms of $use rather than complaining about general issues with $ref.  Otherwise, let's sort out $ref in #66 and then come back to this.

**Comment by handrews** [comment]
@epoberezkin   : What part of "$ref cannot have other keys present because it doesn't merge, it just references" is unclear?  I'm not trying to be snarky here, I really can't figure out how else to say it.  If $ref behaved as you seem to want, it would more or less be $merge and we wouldn't be arguing over that proposal elsewhere (and pleas leave the $merge arguments in issue #15 ).  And in any event, that's not how it behaves, and if you'd like to change its behavior please do that in #66 .

**Comment by handrews** [comment]
I have a whole section about what the existing workarounds are, with "allOf".  It discusses how "allOf"'s behavior with annotations is insufficient for documentation generation, and how "default" simply doesn't work in a meaningful way.

"$ref" is irrelevant- you can't add other keys to it, and using $ref inside of "allOf" vs using a literal schema inside of "allOf" are equivalent for these purposes.

What part of my discussion of the "allOf" alternative is unclear or unconvincing?

**Comment by epoberezkin** [comment]
I will review it again.

Documentation generation is outside of spec. Whether the doc generator can understand $ref wrapped in allOf or not is the problem of the generator - extending the standard create for validation hardly seems an adequate solution.

> "$ref" is irrelevant- you can't add other keys to it, and using $ref inside of "allOf" vs using a literal schema inside of "allOf" are equivalent for these purposes.

Again, please review the cases in the comment above (https://github.com/json-schema-org/json-schema-spec/issues/98#issuecomment-254064538) - $ref and using literal schema are not equivalent.

**Comment by handrews** [comment]
@epoberezkin 
> Documentation generation is outside of spec.

Um.. no.  From the first paragraph of the JSON Schema Core specification:
> JSON Schema is intended to define validation, documentation, hyperlink navigation, and interaction control of JSON data.

While we confirmed in #89 that documenting how multiple schemas interact in an API is out-of-scope, documenting individual schemas is explicitly in-scope.  Note that I don't mean fancy full I18N/L10N end-user documentation, I just mean reading the schema and presenting it at a developer level.

I'm still not going to discuss your concerns about $ref and inclusion here as it distracts from the main proposal.  If you comment on #66, I will address it there.

**Comment by epoberezkin** [comment]
The way it is written, JSON schema itself is intended as a documentation for JSON data. Deriving some other documentation format from the schema (I assume that's what you mean) is something else.

**Comment by handrews** [comment]
> JSON schema itself is intended as a documentation for JSON data

Right.  This is the declared purpose of title and description:

> Both of these keywords can be used to decorate a user interface with information about the data produced by this user interface.

An automated system cannot disambiguate conflicting values for these keywords when trying to figure out what to present in such a user interface.  That makes schemas using these keywords difficult to re-use, which is the problem being addressed here.

Additionally, completely aside from documentation, the other metadata keyword, "default", is unusable in "allOf" without disambiguation.

Please try to address these problems

* Without framing it in terms of $ref (or wait until we resolve that in #66 )
* Without relying on $merge/$patch (or wait until we resolve that in #15 )
* Without ignoring the substance of the proposal because you don't like how I use the word "documentation".  It's fine to not want to support higher-level documentation generation, but focus on the use case _explicitly listed in the specification_ and the problems are still there

**Comment by epoberezkin** [comment]
Ok, I will spend more time to understand this issue, I can be missing something here :). Will come back with the questions. In the meanwhile, I am writing the comment re $refs in #66.

**Comment by tadhgpearson** [comment]
@epoberezkin  - while the allOf + $ref implementation makes theoretical sense to me, it may be difficult to implement simply because  the behavior is not explicity defined in current versions of JSON schema, and causes today's tools to break. 

I think it would be useful to survey the ease of use of both syntaxes with other developers.  I find the $use ( source + with ) syntax to be much clearer - but obviously, that's just my little opinion, and I think it would be good to see what some other users think ;)

**Comment by handrews** [comment]
@epoberezkin now that #66 is resolved, any further thoughts?  This is a high priority for me to resolve one way or the other for the next draft.  The problem it addresses is a major pain point in a current project.

@robbie-mac this addresses your concern about $merge/$patch violating the intent of closing a schema to extension that you raised in #15 .

@seagreen I believe you were also concerned in #15 about $merge/$patch being overly powerful- this proposal essentially restricts $merge down to a form that cannot affect validation.

@awwright any thoughts here?  This was part of the email thread about different re-use use cases that you seemed to like when I first posted it.  Although I'm not sure if that was in response to this particular proposal or one of the other use cases in the thread :-)

And I might as well tag @Relequestual and @jdesrosiers while I'm at it to see if this resonates with anyone else who has been active here recently.

As noted, I'd really like to get this either into the next draft or definitively rejected.

**Comment by epoberezkin** [comment]
I've re-read the thread.

I understand the issue that some processing tools may face. I think that the problem of tooling of any kind should NOT be addressed in the standard that has validation as its primary purpose, as otherwise we will be facing feature creep from users of different tools (e.g. UI generation - why not add something to support them here too? How UI generation is worse than documentation generation? I am not suggesting it, I am using it as an illustration that documentation generation is equally narrow and out-of-scope concern).

I think either of the following is a better option:
1. add $merge/$patch, as it addresses a wider range of problems
2. add nothing and solve schema extension problems (whether for validation or for annotation keywords) to pre-processing / build tools outside of the standard. If this is good instead of $merge, then it is equally good for narrower and less common problems that this $use proposal aims to solve.

Also bear in mind that $use is more complex to define than $merge - it requires maintaining (= writing, extending, discussing, etc.) the list of keywords that are taken into account, while $merge is quite simple - it can be defined once and forever.

**Comment by seagreen** [comment]
@handrews: Thanks for mentioning me here. This would definitely be an improvement over $merge/$patch from my perspective.

That said, I'm actually still against it, but I'm not sure you should take my thoughts into account.

To explain...

My use case is super weird in that I need a way to write JSON schemas that do structural-only validation and absolutely nothing else. I won't even be using `"title"` or `"description"` -- from the perspective of my application that's metadata about the schema, and I already have a standard, separate way to store things like that.

So as you can I'd prefer that the core, structural part of JSON Schema stayed extremely focused on just validation. I think there are reasons this is a good idea -- consider in how many situations things like `"title"` are tracked somewhere other than in the data itself, for instance in filesystems having a `"title"` in the file's data and also having a file name is redundant.

That said, I appreciate that you have pragmatic use cases for both `"title"` and a way to modify it, and that that will probably (and perhaps should) override theoretical/niche considerations about keeping the focus on structural validation.

**Comment by seagreen** [comment]
>  This is a high priority for me to resolve one way or the other for the next draft.

I'd missed this line. Obviously you need to be taken care of one way or another. Would a "JSON Document Schema" similar to the [JSON UI Schema](https://github.com/json-schema-org/json-schema-spec/issues/67) idea be too crazy of a thought? It could include things like "title" and "description", and also a powerful way to override them.

**Comment by handrews** [comment]
> That said, I appreciate that you have pragmatic use cases for both "title" and a way to modify it, and that that will probably (and perhaps should) override theoretical/niche considerations about keeping the focus on structural validation.

`default` is actually the most important, although `title` and `description are important here as well.  But `default` is actually something I use in APIs at runtime even when there's no use for `title` or `description`.

> Would a "JSON Document Schema" similar to the JSON UI Schema idea be too crazy of a thought?

I would not object to such a thing.  Basically splitting the "meta-data" keywords out of validation (where they don't really belong anyway) and expanding that.  I need to be able to easily access the `default` value while working with instances at runtime.  That's the only potentially tricky point I can think of if this is split out.  Do I have to bring in the entirety of JSON Document Schema just to use defaults?  And if so, how easy/hard is it?

**Comment by seagreen** [comment]
> I would not object to such a thing. Basically splitting the "meta-data" keywords out of validation (where they don't really belong anyway) and expanding that.

Oh man, I didn't actually expect anyone to agree with this. This would be so cool!

> I need to be able to easily access the default value while working with instances at runtime. That's the only potentially tricky point I can think of if this is split out. Do I have to bring in the entirety of JSON Document Schema just to use defaults? And if so, how easy/hard is it?

I can't imagine this would be too difficult (obviously handling things like "title" is no trouble at all). I haven't been following all the open issues though, so maybe more complicated documentation-related keywords are being added. If that's the case though it would make having a JSON Schema Document Extension even more important.

**Comment by handrews** [comment]
Another appealing thing about making the documentation stuff an extension is that we could push all of the I18N/L10N stuff there.  I think.  I kind of like the idea of:

JSON Schema core (bare minimum necessary to use any vocabulary)
plus vocabularies for:  validation, documentation, hypermedia, ui generation

Right now validation has just a bit of documentation stuff but there are arguments over what, exactly, it's there for.  This probably calls for an issue of its own- @seagreen do you want to file one for this extension?  I am supportive but have too many other things going on to champion another big proposal.

**Comment by seagreen** [comment]
Done! https://github.com/json-schema-org/json-schema-spec/issues/136

**Comment by handrews** [comment]
As of https://github.com/json-schema-org/json-schema-spec/issues/160#issuecomment-262642122 and https://github.com/json-schema-org/json-schema-spec/issues/158#issuecomment-262351269 I think @epoberezkin I are coming to an agreement that both `$merge`/`$patch` and `$use` belong outside of JSON Schema.

I will be happy to close this as out of scope once #15 (`$merge`/`$patch`) is also closed, since this issue can easily be implemented as an independent preprocessor.

**Comment by tadhgpearson** [comment]
I'm usually using JSON Schema as part of an Open API definition - so I'm fine with this, provided it this extension would be included in the Open API specification. Do you know at what level that is defined?

**Comment by handrews** [comment]
I don't know much about Swagger / Open API's processes, but you could certainly lobby to include it there if it doesn't go into JSON Schema or some extension vocabulary.  I'd use a different prefix than `$` but that would be up to whichever group wants to implement something like this.

It's not entirely trivial to do strictly as a preprocessor step, as you would need to copy and modify things that are otherwise "$ref"'d, and you can get into trouble with circular references that way, but it's plausible.  But if Open API is integrating JSON Schema processing into a larger system, then they could resolve the "use" statements lazily which would make it quite easy.

**Comment by Relequestual** [comment]
I don't think any of the team is in contact with Open API. Would be good to know a point of contact who is willing to discuss things as needed.

**Comment by tadhgpearson** [comment]
@webron or @fehguy are usually a good place to start for Open APIs

**Comment by fehguy** [comment]
I'm all 👂!

**Comment by webron** [comment]
I'm here too. Will have to spend some time reading the ticket. I was also going to reach out regarding some other issue we have with the JSON Schema website (so, less technical).

**Comment by handrews** [comment]
@fehguy @webron this is a documentation feature that might be better implemented outside of JSON Schema proper, and could be implemented as a just-in-time preprocessing step quite easily.  Read the first comment but you probably don't need to go through all of the subsequent discussion as it's more about various objections related to putting it in JSON Schema.

@webron JSON Schema website bugs should be filed at [json-schema-org/json-schema-org.github.io](https://github.com/json-schema-org/json-schema-org.github.io)

**Comment by handrews** [comment]
@fehguy @webron although note also https://github.com/json-schema-org/json-schema-spec/issues/136 which might add a JSON Schema vocabulary related to documentation so this idea could find a home there as well.

**Comment by webron** [comment]
It will probably take more than just reading the first comment, as I haven't followed the changes since draft 4 closely, and I don't know how `$merge` behaves.

**Comment by handrews** [comment]
`$merge` just applies an arbitrary application/merge-patch+json document to a schema.  The problem with that is that it can change validation results in arbitrary ways.  It's like monkeypatching code- occasionally useful or necessary but more for weird test cases than normal usage.

`$use` is an attempt to restrict that idea to annotation keywords without affecting validation.

**Comment by handrews** [comment]
Note that we're pretty close to consensus that `$merge`/`$patch` will **not** go into JSON Schema.

**Comment by epoberezkin** [comment]
@handrews I thought you reached the conclusion that this should not be included as well.

**Comment by handrews** [comment]
@epoberezkin I thought that was already apparent from this issue.

**Comment by awwright** [comment]
If you think there's a better solution you might consider closing this out.

Does #151 seem to describe the problem being targeted here?

A related proposal has been to allow schema templates, or some way to generate a schema based on an argument (I can't think of a compelling one, but maybe allow only a multiple of a number). But usually this can just be done with an "allOf" intersection.

**Comment by handrews** [comment]
@awwright no #151 and #119 are addressing the same problem.  This is strictly a documentation/annotation problem with a much simpler solution.

 I will only close this issue out if #15 is closed.  If there is any possibility of #15 being accepted, this is a much more focused solution and I will argue for this solution instead.

@awwright I'd like to hear your thoughts on whether this issue (`$use`) is in-scope for JSON Schema or not.  That's really the decision to be made (and if it's not in-scope, then I can't see how #15 would be in-scope either).

**Comment by epoberezkin** [comment]
> @epoberezkin I thought that was already apparent from this issue.

just wanted to make sure, probably wasn't reading it attentively enough ;)

I'd be very happy to have all these ideas (including #15, #151, #119) out of scope. They all can be separate I-D. This way we'll keep it simple here.

Once draft 6 is out of the way, we can brainstorm some better ideas (I have none at the moment), but I'm happy to kill them all for now.

**Comment by tadhgpearson** [comment]
JSON Schema forms a major part of the [Open API Spec](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md), which seems to be gathering steam as *the* documentation language for Restful APIs, so I'm very glad to see you all discussing together :grin: 

Specifically, in this example, @handrews proposed _$use_ extension that would be very helpful for implementing Open API specifications. 
For example, today our underlying system returns a generic price object - with amount, currency, etc. that is shared across a wide variety of APIs (Flight Search, Hotel Reservations...). However, [the amounts often have a different functional meaning in each API, which needs to be documented](https://groups.google.com/forum/#!topic/json-schema/YS1dbWXk_4s). In Open APIs 2.0, we can't reuse the same response object for this, because JSON Schema draft 4 requires them all to share the same description. Thus, we end up with a lot of copy-paste implementations of the same underlying object across our specification.

$use addresses this issue by allowing the reuse of all fields of a referenced object, with a few overwrites. If this were supported in the upcoming Open APIs 3.0 spec, it would help us to avoid repeating the same response object over and over in our specs just in order to have different descriptions for just 1 or 2 fields each time. 

However, it sounds like $use will be an extension to JSON Schema, rather than part of the specification itself. If this is the case (@webron @fehguy ) do I need to open a new issue in the Open APIs issue tracker to request it be considered for the Open API 3.0 draft?

**Comment by fehguy** [comment]
@tadhgpearson yes that'd be the right next step. Thank you

**Comment by handrews** [comment]
@tadhgpearson @fehguy It sounds like @awwright might be open to this as a part of JSON Schema, although I can't really tell from his comment so let's see if we can get him to elaborate :-)  I know @epoberezkin is against, but the scenario that @tadhgpearson describes is exactly the problem that I encountered over and over at Riverbed (and why we implemented the too-powerful `$merge` but only ever used it the way I describe `$use`).  I see the same problem in the project that I recently joined.

So this is a widespread problem, and I would argue a much more practical and easily solved problem than the mess over people using `"additionalProperties": false` for error-checking while wanting to also expand the property set during re-use (which is what #151 and #119 are trying to solve).

If there is a lot of evidence that this is needed, I might want to push for it in JSON Schema a bit more.

**Comment by Relequestual** [comment]
@tadhgpearson I personally see Swagger as a branch off from JSON-Schema. I find naming it the Open API Specification somewhat presumptuous. Winning hearts and minds and all that. I'll not say any more or I'll go into full rant mode. Sorry I haven't actually reviewed the issue.

**Comment by epoberezkin** [comment]
@Relequestual lol. It is indeed. I supports some subset of JSON schema spec.

@handrews we are using $merge for actual validation extension here. Having way to extend meta and not actual validation would be very disappointing... And I think both are hacks to be honest. We need to come up with some alternative extension mechanism...

**Comment by tadhgpearson** [comment]
Aha - so @fehguy - I think this issue is already open for consideration in the next iteration of the specs through  https://github.com/OAI/OpenAPI-Specification/issues/556

**Comment by handrews** [comment]
> Having way to extend meta and not actual validation would be very disappointing...

meta isn't "extended", which is why it is addressed differently from validation.  This issue is not about refusing to extend validation, it's about recognizing that the use cases are different.  Meta data needs to be overwritten depending on context- the original values for description or whatever just disappear in the new usage.

Validation rules need to be combined in some way that respects all sources for the combination.  `$use` is not about preventing us from solving that problem it's about splitting off the very different problems related to annotation and solving them in a way that makes sense for them.

**Comment by Relequestual** [comment]
This looks like a reasonable proposal to me. I'm all for this.

**Comment by awwright** [comment]
I'm concerned that this kind keyword -- anything that says "like X, but with Y" -- is too opaque. This makes certain kinds of use cases difficult to reason over.

If I've automatically generated a SQL structure from a JSON Schema, and one of my fields "Y" is a boolean... what happens when I say "Like field Y, but `type:"integer"` instead". However, this isn't any worse than just making a copy of the schema in question.

Adding constraints can already done with "oneOf". So this would only be needed for removing constraints from an existing schema.

What are some use cases for needing to remove a constraint from a schema?

**Comment by handrews** [comment]
@awwright the whole point of `$use` is to make it impossible to do what you are worried about:

> what happens when I say "Like field Y, but type:"integer" instead"? However, this isn't any worse than just making a copy of the schema in question.

`$use` MUST NOT be used with validation keywords.  It is **only** used with annotation keywords (which have no affect on validation).

This issue does not remove (or add) constraints, so use cases for such are not relevant.  That's not what this does.

**Comment by awwright** [comment]
Ok, that's right about my "type" example.

But... I'm still trying to think of a reason "allOf" wouldn't suffice.

This example seems perfectly _valid_ to me:

```javascript
{ allOf: [ {default: []}, {default: null} ] }
```

If that doesn't really make _sense_, the problem is probably not yet fully understanding out what "allOf" means (related: #161).

I haven't studied any implementations that use these keywords, are there any specific ones people here use?

**Comment by handrews** [comment]
I don't think it makes sense, and I don't think it's related to #161 .

If I have that schema, and I want to fill in the default value, do I fill in an empty array or a null?  There is no way to tell.  `$use` provides an unambiguous view without making validation any less predictable.

**Comment by handrews** [comment]
Likewise for title and description.  If I'm using title in a display, which title do I use?  If I'm generating documentation, I could display all of the descriptions but in my experience (and I have done this particular use case a lot), there's one generic schema that is being re-used, that has a generic description.  And then there's a description that is more appropriate for the _specific_ use of that generic schema.  And that is the description that needs to get displayed.  Showing both is just a confusing mess.

**Comment by Relequestual** [comment]
I've marked this as a medium priority issue as it would be useful to OPEN API (or whatever it's called today ;) ) and despite my reservations about Swagger, increasing inter-team collaboration is a great thing, and strengthening those relationships is a good thing to do. 

I think @tadhgpearson explains the use case fairly well at https://github.com/json-schema-org/json-schema-spec/issues/98#issuecomment-264860961

@awwright do you still have reservations regarding this issue? If so, please can you bullet point them?

@handrews Do you want to copy pasta your first comment and make any alterations based on discussions, then replace the first comment with a link to the new comment? (Only a suggestion, if you think it may help move things along, and push people past the previous discussion to focus on moving this issue forward. I feel it's preferable to opening a new issue. I'm easy).

**Comment by handrews** [comment]
@Relequestual I'm not sure that dumping that giant first comment on anyone again will really help.  But here's a summary:

`$use` applies an piece of JSON in its `with` property to the schema in its `source` property, as if the `with` JSON is an `application/merge-patch+json` document.

The difference between `$use` and `$merge` is that `$use` MUST NOT include validation keywords in `with`.

There is a limitation to `application/merge-patch+json` that it is not possible to set a property to the value `null`, as a `null` in JSON Merge Patch means "remove this property".  In practical terms, this is only really an issue for `default`, as it might be useful to have a default of `null`.  If there's agreement on `$use` in general we can discuss what, if anything, to do about this limitation.  

Here is a snippet of meta-schema that shows `source` as a full schema and `with` as a restricted schema.  They hyper-schema meta-schema would probably exclude additional keywords, but I'm not going to write that out here.

```JSON
{
    "definitions": {
        "validationProperties": {
            "enum": ["type", "multipleOf", "maximum", ...]
        },
        "nonValidationSchema": {
            "allOf": [
                {"$ref": "#"},
                {"propertyNames": {"not": {"$ref": "#/definitions/validationProperties"}}}
            ]
        }
    },
    "properties": {
        "$use": {
            "type": "object",
            "properties": {
                "source": {"$ref": "#"},
                "with": {"$ref": "#/definitions/nonValidationSchema"}
            }
        }
    }
}
```

Note that we do not (an never have) had the ability to say what the target of a `$ref` must validate against.  So if someone were to use a `$ref` in the `with` clause (unusual but not unreasonable), the thing referenced must be a nonValidationSchema.  This isn't really a new restriction- currently the thing referenced by `$ref` must be a schema, but we do not have a way to enforce that either.  Issue #141 would actually solve this underspecification, btw, if we think it is a concern.

**Comment by Relequestual** [comment]
OK. I've reviewed this issue and believe we should move forward with it.
@handrews Can you see any outstanding issue or things we didn't have consensus on?
If not, Tagg this as accepted and available! And then we sit back and watch the PRs roll in... right? =D

**Comment by nitmws** [comment]
A gentle follow up, @Relequestual : I've checked [draft-future](https://github.com/json-schema-org/json-schema-spec/milestone/3) and can't find #98 there, and in the schema.json file of the master branch does not show a definition for $use. What progress of this issue may we expect?

**Comment by Relequestual** [comment]
@nitmws Hey there. Thanks for the prompt. There is some administrative work going on for releasing draft-6 right now. Once that's done, we'll regroup and properly scope for draft-7. Strange this isn't part of the milestone though. It's still tagged and as I suggested the tagging system I'll be sure to use it when scoping.

It would be cool if you could include your usecase or the reason your interested in this issue specifically =] Thanks.

**Comment by handrews** [comment]
@nitmws the draft-future milestone isn't really well-curated at the moment.  As @Relequestual says we'll sort it out once draft-6 is out the door.

And I'll add to the request for your use case, even if it is similar to what is already here.  Showing breadth of support helps make the case for adding this.

**Comment by ruifortes** [comment]
On the topic of "Overriding annotation/usage keywords at point of use" this #98 is getting tremendously overcomplicated and in my 50cent opinion it could not be.
Would be nice that before closing #279 which proposes a different approach you could give some feedback why you think that's not a viable solution.
I guess that's why it was closed because discussing yet another approach here seams to just increase confusion.

**Comment by handrews** [comment]
@ruifortes I'm not sure what you mean by "getting tremendously overcomplicated."  This issue has seen a lot of discussion due to high levels of interest in both this proposal and in other related ones.  

But the proposal itself has remained consistent, aside from a few suggestions on more streamlined syntax.  Streamlining the syntax is an easier discussion to have once the approach is agreed upon.  Primarily, this issue is about whether to support the annotation-only approach or not.  The exact syntax is still open to debate.  But preferably only after we decide on the approach.  Before that, the questions about syntax are mostly noise as they may end up moot.

If we can agree on the approach, I'd probably want to re-file this in a new issue to focus exclusively on finding the right syntax.  Having both discussions simultaneously is confusing.

**Comment by epoberezkin** [comment]
I actually prefer a more generic $merge to limited annotation only extension.

But even more, I think they should stay outside of spec as a generic JSON extension mechanisms.

**Comment by handrews** [comment]
@epoberezkin how would a "generic JSON extension mechanism" be indicated in a media type?

**Comment by epoberezkin** [comment]
`application/merge-<anything>+json`, e.g. `application/merge-schema+json` (I assume each subtype needs separate registration, but maybe the whole pattern can be registered?)

Another option is to extend [rfc6839](https://tools.ietf.org/html/rfc6839), but it's probably not as easy.

I can be wrong but at least the first option seems viable.

**Comment by handrews** [comment]
I think the pattern would have to involve extending RFC 6839, but just registering a media type that is "schema, but with merges" is doable.  Annoying, but doable.  Without amending the suffix RFC or something similar, clients can't rely on being able to guess the underlying `<anything>+json` type based on the pattern, but as long as they recognize the whole thing it's workable.

There's still a great deal of support visible in this issue for including the limited `"$use"` form, so I'm not giving up on that just yet.  You're the one who's all about voting and you're clearly outvoted on this one ;-)  But that's not how we're resolving things so this stays open.

**Comment by nitmws** [comment]
To share my use case:

A JSON Schema may have multiple properties referencing the same subschema object - example:

```
...
"properties": {
		"uriProperty1": {
			"description": "Local description of the uriProperty1",
			"title": "Local title of uriProperty1",
			"$ref": "#/definitions/aURI"
		},
		"uriProperty2": {
			"description": "Local description of the uriProperty2",
			"title": "Local title of uriProperty2",
			"$ref": "#/definitions/aURI"
		},
...
```
The referenced subschema may look like that:
```
...
"definitions": {
		"aURI": {
			"title": "Title of aUri",
			"description": "Description of aURI",
			"type": "string",
			"format": "uri"
		}
...
```
The current (draft 5) rule for `$ref` states:  _All other properties in a "$ref" object MUST be ignored._
By that rule it is not possible to apply a local definition to a property referencing a subschema object by $ref, in the example above the title and description of uriProperty1 and uriProperty2 will by ignored and they are replaced by title and definition of the aURI subschema object.
The organisation I'm working for, [IPTC](https://iptc.org), and other scenarios require to provide a local definition of the semantics of a property and a local title even if this property references a data structure defined at a generic, global level. 
The key reason for that need is that it is impossible to write a title and definition for a referenced object covering all the different semantics of properties referencing this object; this leads to a confusion about the semantics. The only practical alternative would be not applying titles and definitions to all involved object but this would downgrade the semantic quality of a schema definition.

**Comment by handrews** [comment]
Thanks, @nitmws, this is helpful!  Particularly the explanation of why you find yourself in this situation.

**Comment by epoberezkin** [comment]
@handrews I think if it gets to a vote we need to present people with alternatives in one place and specifically ask for votes rather than for opinions. Comments hardly count for votes as people don't see all the alternatives, they just comment on a specific idea.

**Comment by handrews** [comment]
@epoberezkin I am against a vote, as it is explicitly discouraged in IETF process recommendations.

We keep linking to the same proposals over and over again, and this is the only one that continues to attract support.  There is strong demand for it and you have not offered a universally supportable alternative (many of us who support this have already registered disapproval of #15), or a compelling  reason why this is such a bad thing.

Your main objection seems to be:

> I think that the problem of tooling of any kind should NOT be addressed in the standard that has validation as its primary purpose, as otherwise we will be facing feature creep from users of different tools

which is not relevant.  These keywords are in the spec, so we need to be able to use them effectively.  I have proposed moving them out of validation to another vocabulary- that's #136 .  I'm more than happy to work that out first, as I dislike having the annotation keywords in the validation vocabulary myself.

But that's not a problem with this proposal, that's a problem with the existing inclusion of metadata in the validation spec.

**Comment by epoberezkin** [comment]
Ok, let's wait untill the draft-6 is out, I'll review all this again.

**Comment by erayd** [comment]
> I am against a vote, as it is explicitly discouraged in IETF process recommendations.

Then for what it's worth, I would like to state my preference in favour of `$merge`. That doesn't necessarily have to be as per #15, but for me inheritance is important, including the ability to to override validation keywords, and `$use` can't do that.

Regarding the reluctance around including `$merge` (or indeed other similar functionality) in the JSON schema spec, this feels quite similar to `$ref` (in that both result in a pre-validation transformation of the schema), and I feel that it should be treated accordingly. If `$ref` is part of the schema spec, then things like `$merge` and `$use` also have a valid claim to be included. If keeping them separate is preferable, then perhaps splitting `$ref` out of the core and into a separate document that defines preprocessing keywords might be a viable solution, and `$merge` / `$use` / etc. can have a logical home there.

Apologies if I'm repeating something that's already been suggested; you guys have written a lot (in many different venues) on this, and I have not read it all.

**Comment by epoberezkin** [comment]
@erayd if we include anything, I'd rather it be $merge - it solves many real issues, and it covers what $use does as well.

I disagree that $ref can be seen as pre-processing. As I've explained in many places you cannot resolve $refs by including target schemas, it's not equivalent.

$merge ono the other hand, is pre-processing.

**Comment by erayd** [comment]
> I disagree that $ref can be seen as pre-processing. As I've explained in many places you cannot resolve $refs by including target schemas, it's not equivalent.

@epoberezkin Are you referring to #66?

**Comment by epoberezkin** [comment]
@erayd yes and it's summarised here: https://github.com/json-schema-org/json-schema-spec/issues/279#issuecomment-288558322

**Comment by erayd** [comment]
@epoberezkin You make good points... having read that, I think I agree with you that `$ref` isn't preprocessing. Thanks for pointing that out :-).

I would also prefer `$merge` over `$use` for standardisation - lacking `$merge` or something equivalent in the standard, I'll just end up preprocessing things to achieve the same behavior *anyway*, so IMO it makes sense that the behavior be standardised, as there will be many others in the same boat. Whether that is documented as part of the schema spec or in another document matters less to me than that it be included *somewhere*.

**Comment by handrews** [comment]
@erayd the argument against `"$merge"` is that you cannot reason about the result at all.  It allows arbitrary transformations, which is not something that we support and is viewed by many as problematic for validation.  Hence treating validation and annotation keywords separately.

**Comment by epoberezkin** [comment]
I'd like us to step back and review again the list of problems that we want to solve - post draft-6 publication. I was very much hoping that some good idea to additionalProperties would come to mind, but only plenty of bad ideas did... Nothing better than $merge so far.

My experience of actually using $merge (I do use it a lot) is that it is easy to reason about it. And that it is difficult to reason about $patch (similar but even more generic idea based on JSON-patch RFC).

Maybe we should require that BOTH parts in $merge should be JSON-schemas and not only source - it is so in all my cases anyway (apart from `null` case to remove the properties) and I not see any reason of merging arbitrary JSON into a schema (that's what would make it difficult indeed). That makes it less generic and still as useful. `null` case is a question, but maybe we should just say that `null` is equivalent to empty schema (in addition to true)?

**Comment by erayd** [comment]
@handrews I did read some of those arguments, and there are good points made, but I disagree that arbitrary transformations are a problem, as long as they're predictable and happen prior to using the transformation result for validation (i.e. as a preprocessing step). Is there any lack of clarity that a more detailed specification wouldn't solve?

It's quite possible that I have missed something important though; there is a lot of content, and I haven't seen all of it. Most of what I read has been in this issue, issue #15, and various comments attached to other things.

> Maybe we should require that BOTH parts in $merge should be JSON-schemas and not only source...

That seems reasonable, noting that the intended outcome is a valid schema that can be used for validation.

**Comment by handrews** [comment]
> I'd like us to step back and review again the list of problems that we want to solve - post draft-6 publication.

Yeah, that's reasonable.

> I was very much hoping that some good idea to additionalProperties problem would come to mind, but only plenty of bad ideas did...

I'm pretty sure there are no good ideas for it.  Personally, I don't care, and am strongly against anything that complicates using annotation keywords in an effort to "solve" problems with `"additionalProperties"` that I don't consider problems (however, that's a personal view, not one that I necessarily expect anyone else to agree with).

If we must "solve" those problems, I want the mechanisms to be separate such that each clearly targets a specific use cases.  Overriding metadata is a totally separate problem from `"additionalProperties"` messes, and the only thing I ever used `"$merge"` for was overriding metadata.  Allowing a validator to completely ignore the effect of `"$use"` is a very nice property for implementation, and it's much easier to get a large team of engineers to use it in the desired manner.

> Maybe we should require that BOTH parts in $merge should be JSON-schemas and not only source

While I'm still against `"$merge"`, that would make it a bit less objectionable.  I still think it's a bad idea.

**Comment by MikeRalphson** [comment]
As requested [here](https://github.com/OAI/OpenAPI-Specification/issues/556#issuecomment-291522328), an analysis of [additional properties](https://github.com/Mermade/openapi-specification-extensions/blob/master/refs/combined.csv) used in combination with `$ref` over 49,366 OpenAPI (Swagger) 2.0 definitions.

**Comment by handrews** [comment]
@MikeRalphson thanks for producing this!

To clarify, by "additional properties" you mean additional properties in the `{"$ref": "#/some/reference"}` object, and _not_ the `"additionalProperties"` JSON Schema keyword, correct?

For the moment, I'm going to ignore keywords that are neither part of JSON Schema nor have an obvious correlation to proposed or about-to-be-published parts of JSON Schema.  I'm also ignoring anything that occurs less than 10 times, to make it a bit more manageable.  With that in mind, here are the main takeaways I see:

78436 involve only non-default annotation keywords (`"title"`, `"description"`, `"examples"`, `"readOnly"`), with `"description"` by far the most common.

64 involve `"default"`, possibly alongside some of the annotation keywords mentioned above

4239 involve `"type"`, usually alongside some of the above

1182 involve `"required"`, in some cases alongside the above, but mostly on its own

Nothing else has significant numbers- the next most significant is `"uniqueItems"` at 82.

-----

Annotation keywords are overridden more than 10x the times of everything else combined.  That, to me, is a pretty compelling argument for `"$use"` (or an equivalent feature with more concise syntax such as just allowing those keywords alongside `"$ref"`) as its own thing.  There is an order of magnitude more demand for this than for keywords that would require `"$merge"`.

As for `"type"`, @MikeRalphson noted elsewhere that in many cases that seems to be adding a type of "object" that was left off of the source, which is possibly a cut-and-paste legacy of some sort.  So there is probably not as much real demand for that as their might appear.

Finally, for `"required"`, I think using `"$merge"` to *remove* a requirement is a particularly bad use of merging/patching.  JSON Schema is a constraint system where constraints are added but never removed.  Introducing constraint-removal functionality is one of the main reasons I object to `"$merge"`.  In a way that is worse than any complaints about possibly producing an invalid schema.  It reverses the entire conceptual model of the system.  That's bad.

This data leaves me more convinced than ever that overriding annotation keywords is a feature that we need, and generic merging/patching is a feature that is neither truly needed nor desirable.

**Comment by MikeRalphson** [comment]
> To clarify, by "additional properties" you mean additional properties in the {"$ref": "#/some/reference"} object, and not the "additionalProperties" JSON Schema keyword, correct?

Correct. The confusing column heading comes from the database table used to hold the results.

**Comment by tadhgpearson** [comment]
Great summary @MikeRalphson! In over 92% of all cases where standard elements accompanied the $ref the author was providing the description. 
With that in mind, maybe you only need to support overriding the description? Maybe the easiest way to express this is, if there's a description accompanying the $ref, it overrides the description in the referred object?

**Comment by handrews** [comment]
**_VOTE-A-RAMA!!!_**

It's time to gauge community support for [all re-use/extension/additionalProperties proposals](https://github.com/json-schema-org/json-schema-spec/issues?q=is%3Aissue+is%3Aopen+label%3A%22re-use+%2F+addlProps%22) and actually make some decisions for the next draft.

Please use emoji reactions **ON THIS COMMENT** to indicate your support.

* You do not need to vote on every proposal
* If you have no opinion, don't vote- that is also useful data
* If you've already commented on this issue, please still vote so we know your *current* thoughts
* Not all proposals solve exactly the same problem, so we may end up implementing more than one

This is **_not_** a binding majority-rule vote, but it will be a very significant input into discussions.

Here are the meanings for the emojis:

* **Celebration**:  I support this so strongly that I want to be the primary advocate for it
* **Heart**:  I think this is an ideal solution
* **Smiley face**: I'd be happy with this solution
* **Thumbs up**: I'd tolerate this solution
* **Thumbs down**: I'd rather we not do this, but it wouldn't be the end of the world
* **Frowny face**: I'd be actively unhappy, and may even consider other technologies instead

If you want to explain in more detail, feel free to add another comment, but *please also vote on this comment*.

The vote should stay open for several weeks- I'll update this comment with specifics once we see how much feedback we are getting and whether any really obvious patterns emerge.

**Comment by nitmws** [comment]
To clarify my position: only the JSON Schema properties title and description defining the semantics of a property (of my schema) should override title and description of a referenced property.
😍 (= Celebration, as this emoji is not available on Github)

**Comment by handrews** [comment]
@nitmws for "Celebration" I meant the party streamer cone thing.  I'll put it on this comment so you can see and vote with it.  I couldn't figure out what else to call it :-P

**Comment by MikeRalphson** [comment]
I believe it's called `:tada:` -> :tada:

**Comment by nitmws** [comment]
Ok, my vote: 🎉

**Comment by Relequestual** [comment]
> My experience of actually using $merge (I do use it a lot) is that it is easy to reason about it. And that it is difficult to reason about $patch (similar but even more generic idea based on JSON-patch RFC). - @epoberezkin 

Spending some time reviewing #15 , is there any reason we couldn't JUST include $merge but not $pathch? $merge seems easier to understand with minimal effort.

I don't know with other languages, but with perl, this concept is great and I utilise it often.

I see general conversation is that people are worried it may be too complicated and result in people creating things they don't mean to, or are invalid. There's always going to be a gap between new users and power users, and right now, power users want a way to do this. While most seem to be mostly interested in just metadata (which $use would cover), I've still seen more than a handful of people ask for something along the lines of $merge, and not just in github issues.

It's always possible to create nonsensical schemas. We can help by providing examples and common pitfalls on the website if we think it would help.


@epoberezkin Can you speak to difficulty to implement? 

@handrews I'm downvoting this because I feel if added to the spec, we will get people asking why $use is limited to non validation key words.

**Comment by erayd** [comment]
> Spending some time reviewing #15 , is there any reason we couldn't JUST include $merge but not $pathch? $merge seems easier to understand with minimal effort.

The null thing, I think. But if it's a choice between having just `$merge` (without `$patch`), or neither of them, then I'd far rather have `$merge`.

**Comment by Relequestual** [comment]
@erayd I thought the null was resolved? https://github.com/json-schema-org/json-schema-spec/issues/15#issuecomment-207262327

@epoberezkin Looks like you were on the same train of thought as me at one point: https://github.com/json-schema-org/json-schema-spec/issues/15#issuecomment-207285175

**Comment by handrews** [comment]
@Relequestual If we do `$merge` I'd like `$patch` as well:  it just makes sense to me to support the two well-defined JSON-based patching media types.  It's extremely difficult to work with arrays in `$merge`, for instance.  I would assume that in most languages, existing libraries for each media type would be used.  I'll comment more in #15 

Looking at my comments in #364, my main reason to push for this was `"readOnly"`.  I've come around to deciding that my application (Doca) can define its own usage for the other keywords.  `"default"` is a bit problematic, but that's true in many more ways than this one.  So if we accept #364 then I actually have no personal need for any  of these solutions.

**Comment by handrews** [comment]
@nitmws , you indicated that you would advocate for this.  Various people are pushing hard for `"$merge"`.  While I'm not a fan of it, if we went that route would that be acceptable for you?  Or do you specifically want to avoid the additional effects of `"$merge"`?

**Comment by erayd** [comment]
@Relequestual I think there are still edge cases with `null` for `enum` and `constant` at least, unless I've missed something?

**Comment by handrews** [comment]
@erayd definitely for `const`, I think that is the only one.  `application/merge-patch+json` does not work with individual array elements, so `enum` can contain `null`s.

**Comment by handrews** [comment]
Also, `default` can be `null`, and `examples`.

**Comment by tadhgpearson** [comment]
I think that some more granularity than the current $ref system is required to prevent massive copy-pasting of fields just to change descriptions. Whether this is directly overriding, $use, or whatever, something is necessary - 🎉

**Comment by handrews** [comment]
@tadhgpearson all of the discussions (elsewhere) about how `"default"` can or should be used has made me re-think the description and title use cases from a similar perspective.

These keywords have no impact on validation.  So there is no implementation/interoperability concern there.

Really, it is application-defined how these metadata/annotation keywords are used.  This is even still true after moving `"readOnly"` over to that section:  neither a validator nor a hyper-client are obligated to do anything about a read-only field.  In fact, given the definition of read-only as really meaning "under the control of the source authority", a hyper-client should not assume that sending a changed read-only field in a request is an error- the authority may prefer to just ignore the attempted change.

So what are the interoperability requirements for these keywords?  They are application-specific.  We may want to guarantee behavior in an API Documentation context, but that would be better handled in the specification for that vocabulary.  That could specify, for instance, that descriptions are concatenated with two newlines between each for display.  Or that a description in a parent schema replaces all in child schemas for the purpose of displaying the parent schema.

UI generation would have similar but perhaps not identical requirements.

And many applications just won't care.

Increasingly, I don't see a need to define combinatoric behavior for the metadata/annotation keywords within the validation or core specification. Application (or application domain)-specific vocabularies would be better places to nail down common behavior.

So for those of us who do not want to allow merging/patching on arbitrary keywords, there's not actually anything that needs to be done.  Except maybe opening an issue on the API Doc vocabulary at https://github.com/json-schema-org/json-schema-vocabularies/issues/new

**Comment by nitmws** [comment]
Re https://github.com/json-schema-org/json-schema-spec/issues/98#issuecomment-326424661 : hm, I would have to dig into all the issues regarding `$match` first ... Currently my available time is limited.

**Comment by handrews** [comment]
@nitmws current discussions with @Relequestual and others are leaning towards punting all of this out of draft-07 (to get it wrapped up in the next month) and making it the primary focus of draft-08 (which would be published as soon as we have a resolution, possibly much sooner than 6 months after draft-07).  When we've decided one way or another I'll update all the issues.  Some will get closed as non-viable, while others will probably get consolidated into a few issues summarizing these epic discussions.  No one wants to read 100 comments, many of which go in circles anyway :-P

**Comment by tadhgpearson** [comment]
@handrews  - I think that it's very important to specify this.

As a user, it's a major advantage to be able to use JSON schema with many tools for many purposes. I can use it for a documentation generator on my website, a code generator and publish it in 3rd party directories which might have their own UIs. If the behavior of, for example `description` is not specified, then the tools may behave differently, and I can no longer effectively guarantee the format of my descriptions in generated documentation.

This isn't an arbitrary problem. 
In early versions of Swagger 2 the `default` keyword was left open to interpretation. As a result, `default` means _example value_ in Smartdocs (which was used for documentation generation on a website) but _default value if no input is provided_ in some code generators. (This is now the official meaning in Swagger.) 
The result is users in India complained that generated code automatically did a flight search from Boston to London if the user provided no input, simply because of this default value for the documentation - which is obviously not ideal behavior!

**Comment by handrews** [comment]
@tadhgpearson I think I wasn't clear.  I think specifying makes sense, I'm just not convinced it should be specified in the _validation_ spec.  There is a larger discussion going on about whether the whole "Meta-data keywords" section belongs in the validation spec at all.  That won't change in draft-07 but it is being kicked around, in part for exactly this reason: the community wants these things specified more, but such specification isn't really about data validation.

By moving it to a separate (but still "official", whatever that means) specification, we can address the real concerns around these keywords without limiting validation behavior.  All of your examples are about generating docs, code or UI.  That's what the new specification(s) would focus on.

None of your examples have anything to do with validation, unless I'm missing something.

What we've come to realize is that validation and data definition are not the same thing, and a lot of the difficulty various tools struggle with has to do with finding a way to support concepts that are really only useful for validation.  Or dealing with people complaining that they are not supported, even when the use cases for data definition are unclear.

**Comment by erayd** [comment]
@tadhgpearson
That's not really related to `$use` though. That's more an issue of `default` being ambiguously specified, resulting in schema authors using it to mean several different things. If `default` is properly specified and / or split into multiple keywords, then the problem goes away.

The same comment applies to other metadata keywords, although I think `default` is the most severe example.

**Comment by tadhgpearson** [comment]
@erayd Correct, absolutely not related to `$use`.
To clarify - my example above was related to the `default` in Swagger 2 input parameters, and was not related to JSON-schemas `default`. I was just using it as a way to highlight the issues that occur when field usage is left without a clear specification.

**Comment by handrews** [comment]
@erayd yup.  We toyed with the multiple keywords approach for `default`, but it bogged down.  I think that was another side effect of trying to specify meta-data keywords like validation keywords.

For data definition, I think the specifications end up being a bit looser.  We need to avoid things like accidentally using examples as runtime defaults (`examples` obviously helps with that).  But when combining schemas with meta-data keywords, there may be a number of reasonable strategies that different applications might choose.  We can then recommend a range of options as long as they don't actively clash with each other.  Or specify things like `default` MAY be used as a runtime default inserted into data, but such usage MUST be controlled by user configuration.  Or something like that- I just made that up, don't take it too seriously.

**Comment by erayd** [comment]
@handrews
Mmm, there's a whole big discussion to be had there I think. Ambiguity around `default` is the issue that motivated me to get involved with the spec in the first place. Draft-09 perhaps?

**Comment by handrews** [comment]
@erayd it will probably go over to json-schema-org/json-schema-vocabularies.  Which is currently set up for three vocabularies for (docs, UI, code) generation.  I'm starting to think that there is actually a basic data definition spec to work out, which may work for some and form the basis for others (obviously annotating whether an enum should be a drop-down or a set of radio buttons is UI-specific).

That would also let it evolve independent of the existing schedule.  I expect it would be driven by a slightly different set of participants, just as we have people who look primarily at hyper-schema, and people who only care about validation.

**Comment by handrews** [comment]
@erayd @tadhgpearson we actually already have an initial proposal from someone, which you can see here:
https://github.com/json-schema-org/json-schema-vocabularies/pull/4

there's some discussion already, and we'll pick it up more after draft-07 (but feel free to comment)

**Comment by erayd** [comment]
@handrews I like the sound of that.

**Comment by handrews** [comment]
At the end of the vote-a-rama, I said that I would consolidate these issues to focus the discussion for draft-08.  I've filed #515 which outlines the two possible competing approaches.  It's pretty abstract, but once we choose an approach, deciding which exact keyword and behavior we want should be less controversial.  Therefore I'm closing this in favor of #515.
